Homework 03 PR Sasha3211

Sasha2511 opened the pull request:
This is a PR with my solutions for Homework 03

Sasha2511 committed:
ArithmeticMeanCalculator resolving

Sasha2511 committed:
CelsiusToFahrenheitConvertor resolving

Sasha2511 committed:
DivisionWithRemainderCalculator resolving

Sasha2511 committed:
HoursToMinutesConvertor resolving

Sasha2511 committed:
calculatePercent resolving
also added calculation from example 8% from 50

Sasha2511 committed:
Salary Calculation resolving
Used 38.1575 Currency rate for USD. Can be changed

vikmaksymenko commented on src/main/java/part1/ArithmeticMeanCalculator.java:
```diff
@@ -2,8 +2,7 @@
     public static double calculateArithmeticMean(double a, double b, double c) {
-        // TODO: Implement solution here
-        return -1;
+        return (a + b + c)/3;
```
Please, update formatting for solutions. Use https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html#formatter-in-settings

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part1/CelsiusToFahrenheitConvertor.java:
```diff
@@ -2,8 +2,8 @@
     public static double convertCelsiusToFahrenheit(double celsius) {
-        // TODO: Implement solution here
-        return -1;
+        double farenheit = celsius * (9.0 / 5.0) +32.0;
```
`32.0` can be simply `32`, java will cast it to double

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part1/SalaryCalculator.java:
```diff
@@ -1,14 +1,24 @@
-        // TODO: Implement solution here
-        return -1;
+        double currencyRateUSD = 38.1575;
+        double hourlyRateHRN = hourlyRate * currencyRateUSD;
```
For hryvnias, you should use `UAH` 
Also, you don't need to change the currency, we consider, that the salary is already in hryvnias

Sasha2511 replied:
Understood

vikmaksymenko commented on src/main/java/part1/SalaryCalculator.java:
```diff
@@ -1,14 +1,24 @@
+        double currencyRateUSD = 38.1575;
+        double hourlyRateHRN = hourlyRate * currencyRateUSD;
+        double resultBT = hoursPerMonth * hourlyRateHRN;
+        double eSV = 1562.0;
```
`currencyRateUSD`, `percent` and  `eSV` can be constants

Sasha2511 replied:
fixed

vikmaksymenko reviewed (changes requested):
Please, merge latest changes from `main` to your branch and make sure, that the tests are passing

Sasha2511 committed:
ArithmeticMeanCalculator resolving
Code formatting is applied in accordance with comment in PR

Sasha2511 committed:
CelsiusToFahrenheitConvertor resolving
32.0 changed to 32 in accordance with comment in PR

Sasha2511 committed:
SalaryCalculator resolving
Deleted currency rates, applied final to eSV and percent in accordance with comment in PR

Sasha2511 committed:
CelsiusToFahrenheitConvertor resolving
Checked this solution in keys branch. Made exactly the same.

Sasha2511 committed:
needed commit before (or after)? mergim main

Sasha2511 committed:
SalaryCalculator test fix 01

Sasha2511 committed:
EvenChecker resolving

Sasha2511 committed:
MaxOfThreeChecker resolving

Sasha2511 committed:
LeapYearChecker resolving

Sasha2511 committed:
VowelChecker resolving

Sasha2511 committed:
Merge branch 'main' into sasha2511

Sasha2511 committed:
convertPointsToMarks resolving

Sasha2511 committed:
TemperatureConvertor resolving

Sasha2511 committed:
TemperatureConvertor resolving with auto formatted code

Sasha2511 committed:
TemperatureConvertor resolving with auto formatted code 02

Sasha2511 committed:
TemperatureConvertor resolving with auto formatted code 03

Sasha2511 committed:
DartsGame resolving

Sasha2511 committed:
AgeOnPlanetCalculator resolving

Sasha2511 committed:
AgeOnPlanetCalculator resolving with some changes in variable types

Sasha2511 committed:
MathCalculator first 3 items resolved, to continue

Sasha2511 committed:
MathCalculator first 3 items resolved, fixed isPrime calculation, to continue

Sasha2511 committed:
MathCalculator first 3 items resolved, fixed isPrime calculation iteration 02, to continue

Sasha2511 committed:
printFibonacciNumbers resolving

Sasha2511 committed:
printFibonacciNumbers resolving fixing calculation

Sasha2511 committed:
printFibonacciNumbers resolving fixing calculation fix 02

Sasha2511 committed:
printDigitsReversed resolving and improving printFibonacciNumbers

Sasha2511 committed:
isArmstrongNumber resolving

Sasha2511 committed:
collatzConjecture resolving

Sasha2511 committed:
collatzConjecture resolving fixed calculation

Sasha2511 committed:
DepositCalculator resolving

vikmaksymenko commented on src/main/java/part3/DepositCalculator.java:
```diff
@@ -2,7 +2,14 @@
+        int i = 1;
+        double a = principal * (1 + rate);
+        System.out.println(i + " year Deposit is: " + a);
+        while (i < years) {
```
Краще використати цикл for

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part3/MathCalculator.java:
```diff
@@ -2,35 +2,88 @@
+            return valueOfFactorial;
+        } else {
+            int i = 1;
+            while (i <= n) {
```
Так само, for  би підійшов краще

vikmaksymenko replied:
Фактично, весь цей блок зведеться до 
```java
        int valueOfFactorial = 1;
        for(int i = 1; i <= n; i++) {
            valueOfFactorial *= i;
        }
```

Sasha2511 replied:
fixed

Sasha2511 committed:
random, max, min, avg, reverse, indexOf, removeDuplicates resolving

vikmaksymenko commented on src/main/java/part2/EvenChecker.java:
```diff
@@ -2,9 +2,7 @@
-        // For sake of simplicity, we assume that 0 is an even number
-        // TODO: Implement solution here
-        return false;
+        return (number % 2) == 0;
```
Дужки не обов'язкові, але ок

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part2/LeapYearChecker.java:
```diff
@@ -4,7 +4,13 @@ public class LeapYearChecker {
     public static boolean isLeapYear(int year) {
         // Read about leap years here: https://en.wikipedia.org/wiki/Leap_year
         // TODO: Implement solution here
```
Цей TODO вже не треба, прибирайте зайві коментариі :)

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part2/LeapYearChecker.java:
```diff
@@ -4,7 +4,13 @@ public class LeapYearChecker {
         // Read about leap years here: https://en.wikipedia.org/wiki/Leap_year
         // TODO: Implement solution here
-        return false;
+        boolean isLeapYear;
+        if (year % 100 == 0 && year % 400 != 0) {
+            isLeapYear = false;
+        } else {
+            isLeapYear = year % 4 == 0;
+        }
+        return isLeapYear;
```
Опціонально: цей код можна трохи спростити 
```java
        if (year % 100 == 0 && year % 400 != 0) {
            return false;
        }

        return year % 4 == 0;
```

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part2/PointsToMarksConvertor.java:
```diff
@@ -3,8 +3,15 @@
-        return 'X';
+        char mark;
+
+        if (points < 60) mark = 'F';
+        else if (points <= 74) mark = 'D';
+        else if (points <= 84) mark = 'C';
+        else if (points <= 89) mark = 'B';
+        else mark = 'A';
```
Я б радив використати варіант з ключів: 
```java
        if (points < 60) {
            return 'F';
        } else if (points < 75) {
            return 'D';
        } else if (points < 85) {
            return 'C';
        } else if (points < 90) {
            return 'B';
        } else {
            return 'A';
        }
```
Ваш варіант правильний, але є один нюанс. У випадку, якщо замість цілих чисел в якості параметру будуть передаватись числа з плаваючою комою, умови доведеться оновити. Бо тоді наприклад 89.5 буде дорівнювати А

vikmaksymenko replied:
Перечитав код з ключів, зрозумів, що в ньому не потрібний else-if 🤦 Буває

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part3/MathCalculator.java:
```diff
@@ -2,35 +2,88 @@
     public static boolean isArmstrongNumber(int n) {
-        // TODO: Implement solution here
-        return false;
+        int degree = String.valueOf(n).length();
```
Переводити число в рядок, щоб порахувати кількість цифр - це вже якось занадто. Воно звісно працює, але давайте спробуємо зробити те саме за допомогою базових операторів

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part3/MathCalculator.java:
```diff
@@ -2,35 +2,88 @@
+
+        while (num > 0) {
+            digit = num % 10;
+            k = (int) (k + Math.pow(digit, degree));
```
Назва змінної k не дуже очевидна, подумайте над варіантом, який краще описує її призначення

Sasha2511 replied:
fixed

vikmaksymenko commented on src/main/java/part3/MathCalculator.java:
```diff
@@ -2,35 +2,88 @@
-        // TODO: Implement solution here
+        int a = 0;
+        int b = 1;
+        if (n == 1) {
+            System.out.println(a);
+            return;
+        }
+        System.out.println(a);
```
Зверніть увагу на цей фрагмент коду. Фактично, ви виводите `a` незаложено від умови. Тому краще переформатувати код наступним чином: 
```java
            System.out.println(a);
           if (n == 1) {
                return;
           }
```

Sasha2511 replied:
fixed

Sasha2511 committed:
random, max, min, avg, reverse, indexOf, removeDuplicates resolving also resolving merge conflict

Sasha2511 committed:
Deposit Calculator fixed

Sasha2511 committed:
InArrayOperator before changes

Sasha2511 committed:
InArrayOperator after changes
With resolving fillWithNumbers,max, min, avg, reverse, copy, indexOf, removeDuplicates

Sasha2511 committed:
InArrayOperator again before changes

Sasha2511 committed:
InArrayOperator with resolved fillWithNumbers, max, min, avg, reverse, copy, indexOf, removeDuplicates, added some code to check solution

Sasha2511 committed:
InArrayOperator with resolved fillWithNumbers, max, min, avg, reverse, copy, indexOf, removeDuplicates, added some code to check solution 01

Sasha2511 committed:
Some fixes in MathCalculator and additional changes in IntArrayOperations

Sasha2511 committed:
Merge branch 'main' into sasha2511

Sasha2511 committed:
MathCalculator, PointsToMarksConvertor, LeapYearChecker, EvenChecker fixes

Sasha2511 committed:
IntArrayOperations merge resolving

Sasha2511 committed:
IntArrayOperations sorting resolving (looked in keys)

Sasha2511 committed:
IntArrayOperations isSorting resolving

Sasha2511 committed:
IntArrayOperations sum resolving

Sasha2511 committed:
binarySearch resolving

Sasha2511 committed:
Fixes to pass tests

Sasha2511 committed:
getRandomCard resolving

Sasha2511 committed:
Matrix resolving

Sasha2511 committed:
reduceMatrixToArray resolving

vikmaksymenko commented on src/main/java/part4/IntArrayOperations.java:
```diff
@@ -2,67 +2,253 @@
     public static int avg(int[] array) {
-        // TODO: Implement solution here
-        return -1;
+        if (array.length == 0) {
+            return 0;
+        }
+        int numberAvg = 0;
+        for (int j : array) {
+            numberAvg += j;
+        }
```
Ви могли викликати метод sum, щоб не дублювати логіку

Sasha2511 replied:
Викликав

vikmaksymenko commented on src/main/java/part4/IntArrayOperations.java:
```diff
@@ -2,67 +2,253 @@
     public static int[] reverse(int[] array) {
-        // TODO: Implement solution here
-        return null;
+        int[] reversedArray = new int[array.length];
+        for (int i = 0, j = (array.length - 1); i < array.length; i++, j--) {
+            reversedArray[j] = array[i];
+        }
+        return reversedArray;
```
Подумайте, чи можна в цій задачі обійтися одним індексом?

Sasha2511 replied:
Зробив

vikmaksymenko commented on src/main/java/part4/IntArrayOperations.java:
```diff
@@ -2,67 +2,253 @@
-        // TODO: Implement solution here
-        return null;
+        int[] withNoDuplicates = new int[array.length];
+        int i = 0;
```
Краще змінну `i` залишити для назви індексів. Ба більше, я б не радив використовувати i/j для елементів в for-each, бо вони здебільшого ассоціюються з індексами в звичайному for

Sasha2511 replied:
Зробив

vikmaksymenko commented on src/main/java/part4/IntArrayOperations.java:
```diff
@@ -2,67 +2,253 @@
     public static int[] sort(int[] array, boolean ascending) {
-        // TODO: Implement solution here
-        return null;
+        int[] sorted = copy(array);
+        for (int i = 0; i < sorted.length; i++) {
+            for (int j = i + 1; j < sorted.length; j++) {
+                if (ascending) {
+                    if (sorted[i] > sorted[j]) {
+                        swap(sorted, i, j);
+                    }
+                } else {
+                    if (sorted[i] < sorted[j]) {
+                        swap(sorted, i, j);
+                    }
+                }
+            }
+        }
+        return sorted;
```
Який алгоритм сортування ви використали?

Sasha2511 replied:
Сортування вибором

vikmaksymenko commented on src/main/java/part4/IntArrayOperations.java:
```diff
@@ -2,67 +2,253 @@
+
+        for (int k = array1.length; k < lengthOfMergedArray; k++) {
+            mergedArray[k] = array2[n];
+            n++;
```
n++ можна винести в ітератор всередині for. Але я б краще залишив його як є, але в другому циклі теж просто проходив по всім елементам

Sasha2511 replied:
Переписав

Sasha2511 committed:
IntArrayOperations fixes after comments
